### DISCUSSION TITLE
normals vs mapped normals

### DISCUSSION THREAD

**Discussion by gabburgio** [idaholab/moose#25000] [Q&A General]
Hello all

I am trying to implement an array boundary condition to model albedo from a reflector in multigroup diffusion theory.
This is my current attempt at the class:

source

```
#include "ArrayAlbedoBC.h"

registerMooseObject("pertApp", ArrayAlbedoBC);

InputParameters
ArrayAlbedoBC::validParams()
{
  InputParameters params = ArrayIntegratedBC::validParams();

  params.addRequiredParam<RealEigenMatrix>("albedo_matrix","(incoming current in group i)/(outgoing current in group j)");
  params.addRequiredParam<MaterialPropertyName>("diffusivity", "____");

  return params;
}

ArrayAlbedoBC::ArrayAlbedoBC(const InputParameters & parameters)
  : ArrayIntegratedBC(parameters),
    _array_normals(_assembly.mappedNormals()),
    _grad_u(_var.gradSln()),
    _diffusivity(getMaterialProperty<RealEigenVector>("diffusivity")),
    _albedo_matrix(getParam<RealEigenMatrix>("albedo_matrix"))

{
}


void
ArrayAlbedoBC::computeQpResidual(RealEigenVector & residual)
{
  auto one = Eigen::MatrixXd::Identity(_diffusivity.size(),_diffusivity.size());

  //source = (A - Id)* (1/4 Phi - 1/2 D dPhi / dn)  cfr. Stacey 76

  residual = -((_albedo_matrix - one) * ( 0.25*_u[_qp] - 0.5* (_diffusivity[_qp].asDiagonal() * _grad_u[_qp]) * array_normals[_qp]) ) * _test[_i][_qp];

}

```

header:
```

#pragma once

#include "ArrayIntegratedBC.h"

class ArrayAlbedoBC : public ArrayIntegratedBC 
{
public:
  static InputParameters validParams();

  ArrayAlbedoBC(const InputParameters & parameters);

protected:
  virtual void computeQpResidual(RealEigenVector & residual) override;


   const MaterialProperty<RealEigenVector> & _diffusivity;    

   const ArrayVariableGradient & _grad_u;
   const RealEigenMatrix & _albedo_matrix; 

  /// Normals in type of Eigen map
  const std::vector<Eigen::Map<RealDIMValue>> & _array_normals;

};

```

This seems to work for a four group calculation but gives me a segfault for a 2 group one.

I have also tried to use the normals from arrayintegratedBC (which I am extending) but I get type errors from eigen during compilation:

```
In member function 'virtual void ArrayAlbedoBC::computeQpResidual(libMesh::RealEigenVector&)':
/home/gburgio/projects/pert/src/bcs/ArrayAlbedoBC.C:34:111: error: no match for 'operator*' (operand types are 'const Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<double, double>, const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, const Eigen::Matrix<double, -1, 3> >, const Eigen::Product<Eigen::DiagonalWrapper<const Eigen::Matrix<double, -1, 1> >, Eigen::Matrix<double, -1, 3>, 1> >' and 'const libMesh::Point')
   34 |   residual = -((_albedo_matrix - one) * ( 0.25*_u[_qp] - 0.5* (_diffusivity[_qp].asDiagonal() * _grad_u[_qp]) * _normals[_qp]) ) * _test[_i][_qp];

```

Any advice on how to fix this is greatly appreciated

**Comment by lindsayad** [comment]
@zachmprince @YaqiWang any idea how to help?

**Comment by gabburgio** [comment]
Ok I think I figured it out, dumb mistake on my part in using 

`  auto one = Eigen::MatrixXd::Identity(_diffusivity.size(),_diffusivity.size());`

instead of:

`  auto one = Eigen::MatrixXd::Identity(_diffusivity[_qp].size(),_diffusivity[_qp].size());`

(I realize there's surely a better way to do this)

I would still like to ask whether using the mapped normals is indeed the correct choice here, and, if so, why.
Is it just to have access to Eigen linear algebra methods?

Thanks

**Comment by gabburgio** [comment]
I would also like to ask whether 

`    _grad_u(_var.gradSln()),`

is the right way to retrieve the gradient, or if for instance it will give me problems with coupled simulations and I should try to do something like

    `_grad_u(coupledArrayGradient("variable")),`

with "variable" coming from ResidualObject.h (although I found that doesn't work, perhaps I should set up an interface)

**Reply by GiudGiud** [reply]
GradSln is fine, it s just a tad more complicated than using coupledGradient

**Reply by lindsayad** [reply]
There are also some nuances if you're doing explicit calculations, e.g. you may want not the current gradient. `coupledGradient` takes the implicit/explicit status of a kernel into account

**Reply by GiudGiud** [reply]
also coupledArrayGradient is for array variable gradients. And coupledGradient for regular variables. I think you have it right
